File: ios/Classes/Tasks/OfflineMap/GenerateOfflineMapJobController.swift
//
// Created by Valentin Grigorean on 24.06.2022.
//

import Foundation
import ArcGIS

class GenerateOfflineMapJobController: NSObject {
    private let channel: FlutterMethodChannel
    private let offlineMapJob: AGSGenerateOfflineMapJob
    var status: AGSJobStatus = .notStarted

    init(messenger: FlutterBinaryMessenger, id: Int, offlineMapJob: AGSGenerateOfflineMapJob) {
        self.offlineMapJob = offlineMapJob
        channel = FlutterMethodChannel(name: "plugins.flutter.io/arcgis_channel/offline_map_task/job_\(id)", binaryMessenger: messenger)
        super.init()
        channel.setMethodCallHandler(handle)
        offlineMapJob.progress.addObserver(self, forKeyPath: "fractionCompleted", options: [.old, .new], context: nil)
    }

    deinit {
        offlineMapJob.progress.removeObserver(self, forKeyPath: "fractionCompleted")
        channel.setMethodCallHandler(nil)
    }

    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey: Any]?, context: UnsafeMutableRawPointer?) {

        switch (keyPath) {
        case "fractionCompleted":
            let fractionCompleted = offlineMapJob.progress.fractionCompleted
            channel.invokeMethod("onProgressChanged", arguments: fractionCompleted)
            break
        default:
            break
        }
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getJobProgress":
            result(offlineMapJob.progress.fractionCompleted)
            break
        case "getJobStatus":
            result(offlineMapJob.status.rawValue)
            break
        case "startJob":
            offlineMapJob.start(statusHandler: { [weak self]status in
                guard let self = self else {
                    return
                }
                if status == self.status {
                    return
                }
                self.status = status
                self.channel.invokeMethod("onStatusChanged", arguments: status.rawValue)
            }, completion: { [weak self]result, error in
                NSLog("\(String(describing: error))")
            })
            result(true)
            break
        case "cancelJob":
            offlineMapJob.progress.cancel()
            result(true)
            break
        case "pauseJob":
            offlineMapJob.progress.pause()
            result(true)
            break
        case "getError":
            result(offlineMapJob.error?.toJSON())
            break
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/tasks/offlinemap/GenerateOfflineMapJobController.java
package com.valentingrigorean.arcgis_maps_flutter.tasks.offlinemap;

import androidx.annotation.NonNull;

import com.esri.arcgisruntime.concurrent.Job;
import com.esri.arcgisruntime.tasks.offlinemap.GenerateOfflineMapJob;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class GenerateOfflineMapJobController implements MethodChannel.MethodCallHandler {
    private final MethodChannel channel;
    private final GenerateOfflineMapJob offlineMapJob;
    private final Runnable onProgressChangedRunnable;
    private final Runnable onJobChangedRunnable;
    private Job.Status status = Job.Status.NOT_STARTED;

    public GenerateOfflineMapJobController(BinaryMessenger messenger, int id, GenerateOfflineMapJob offlineMapJob) {
        this.channel = new MethodChannel(messenger, "plugins.flutter.io/arcgis_channel/offline_map_task/job_" + id);
        this.offlineMapJob = offlineMapJob;
        channel.setMethodCallHandler(this);
        onProgressChangedRunnable = () -> {
            channel.invokeMethod("onProgressChanged", offlineMapJob.getProgress() / 100);
        };
        onJobChangedRunnable = () -> {
            final Job.Status jobStatus = offlineMapJob.getStatus();
            if (jobStatus != status) {
                status = jobStatus;
                channel.invokeMethod("onStatusChanged", status.ordinal());
            }
        };
        offlineMapJob.addProgressChangedListener(onProgressChangedRunnable);
        offlineMapJob.addJobChangedListener(onJobChangedRunnable);
    }

    public void dispose() {
        offlineMapJob.removeProgressChangedListener(onProgressChangedRunnable);
        offlineMapJob.removeJobChangedListener(onJobChangedRunnable);
        channel.setMethodCallHandler(null);
    }

    @Override
    protected void finalize() throws Throwable {
        dispose();
        super.finalize();
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        switch (call.method) {
            case "getJobProgress":
                result.success(offlineMapJob.getProgress() / 100);
                break;
            case "getJobStatus":
                result.success(offlineMapJob.getStatus().ordinal());
                break;
            case "startJob":
                offlineMapJob.start();
                break;
            case "cancelJob":
                result.success(offlineMapJob.cancel());
                break;
            case "pauseJob":
                result.success(offlineMapJob.pause());
                break;
            default:
                result.notImplemented();
                break;
        }
    }
}
